src/NavierStokes2D.hpp

```diff
@@ -0,0 +1,633 @@
+  };
+
+  // Block-triangular preconditioner.
+  class PreconditionBlockTriangular
+  {
+  public:
+    // Initialize the preconditioner, given the velocity stiffness matrix, the
+    // pressure mass matrix.
+    void
+    initialize(const TrilinosWrappers::SparseMatrix &velocity_stiffness_,
+               const TrilinosWrappers::SparseMatrix &pressure_mass_,
+               const TrilinosWrappers::SparseMatrix &B_)
+    {
+      velocity_stiffness = &velocity_stiffness_;
+      pressure_mass = &pressure_mass_;
+      B = &B_;
+
+      preconditioner_velocity.initialize(velocity_stiffness_);
+      preconditioner_pressure.initialize(pressure_mass_);
+    }
+
+    // Application of the preconditioner.
+    void
+    vmult(TrilinosWrappers::MPI::BlockVector &dst,
+          const TrilinosWrappers::MPI::BlockVector &src) const
+    {
+      SolverControl solver_control_velocity(10000,
+                                            1e-2 /** src.block(0).l2_norm()*/);
+      SolverCG<TrilinosWrappers::MPI::Vector> solver_cg_velocity(
+          solver_control_velocity);
+      solver_cg_velocity.solve(*velocity_stiffness,
+                               dst.block(0),
+                               src.block(0),
+                               preconditioner_velocity);
+
+      tmp.reinit(src.block(1));
+      B->vmult(tmp, dst.block(0));
+      tmp.sadd(-1.0, src.block(1));
+
+      SolverControl solver_control_pressure(10000,
+                                            1e-2 /* * src.block(1).l2_norm()*/);
+      SolverCG<TrilinosWrappers::MPI::Vector> solver_cg_pressure(
+          solver_control_pressure);
+      solver_cg_pressure.solve(*pressure_mass,
+                               dst.block(1),
+                               tmp,
+                               preconditioner_pressure);
+    }
+
+  protected:
+    // Velocity stiffness matrix.
+    const TrilinosWrappers::SparseMatrix *velocity_stiffness;
+
+    // Preconditioner used for the velocity block.
+    TrilinosWrappers::PreconditionILU preconditioner_velocity;
+
+    // Pressure mass matrix.
+    const TrilinosWrappers::SparseMatrix *pressure_mass;
+
+    // Preconditioner used for the pressure block.
+    TrilinosWrappers::PreconditionILU preconditioner_pressure;
+
+    // B matrix.
+    const TrilinosWrappers::SparseMatrix *B;
+
+    // Temporary vector.
+    mutable TrilinosWrappers::MPI::Vector tmp;
+  };
+  class PreconditionSIMPLE
+  {
+  public:
+    void
+    initialize(const TrilinosWrappers::SparseMatrix &F_,
+               const TrilinosWrappers::SparseMatrix &B_,
+               const TrilinosWrappers::SparseMatrix &B_t,
+               const TrilinosWrappers::MPI::BlockVector &sol_owned)
+    {
+      F = &F_;
+      B = &B_;
+      B_T = &B_t;
+
+      diag_D_inv.reinit(sol_owned.block(0));
+
+      for (unsigned int i : diag_D_inv.locally_owned_elements())
+      {
+        double temp = F->diag_element(i);
```

Comment:
Whenever you know that a variable is not supposed to change, you should declare it `const`:
```suggestion
        const double temp = F->diag_element(i);
```
However, I think that computing the inverse in one line would probably work just as fine here:
```cpp
diag_D_inv[i] = 1.0 / F->diag_element(i);
```